contracts/DataAvailability.sol

```diff
@@ -0,0 +1,223 @@
+     * across timestamps.
+     * 
+     * The contract owner and any permitted uploaders may upload user objective function data
+     * at a given timestamp. The objective function data that are uploaded represent the delta
+     * between the objective function value at the current timestamp and the value at the
+     * previous timestamp that data was uploaded. The data is not actually stored in the contract,
+     * but events are emitted for each upload, which can be used to reconstruct the full data history.
+     *
+     * Instead of storing all data in the contract, a rolling hash of the data is calculated and stored
+     * for each timestamp. When verifying the data, the submitted proof will contain a hash of the data
+     * computed in a trusted zkVM; if the hash contained in the proof matches the hash stored in the contract,
+     * the data is considered valid.
+
+     * The rolling hash is calculated by summing the existing hash for a given timestamp with the hash of each
+     * user:value pair modulo 2^256 as they are received by the contract. Since modular addition is both commutative and associative, this
+     * results in a hash that is invariant to the order of uploads. Since we cannot "remove" information about
+     * a data entry from the hash, we require that information about a user:value pair is only submitted once
+     * per timestamp, to ensure that the hash is correct.
+     *
+     * The intended usage pattern of this contract is as follows:
+     * - For a given timestamp t_1, changes in objective function values since the last timestamp t_0 are calculated.
+     * - These data are submitted to the contract, and a rolling hash is calculated.
+     * - The contract emits events for each upload, which can be used to reconstruct the full data history.
+     * - These data are used by an off-chain reward distribution service to calculate rewards owed for the
+     *   period between t_0 and t_1.
+     */
+
+  // Role for authorized uploaders
+  bytes32 public constant UPLOADER_ROLE = keccak256('UPLOADER_ROLE');
+
+  // Mapping to store the rolling hash for each timestamp
+  mapping(uint256 => bytes32) private timestampHashes;
+
+  // Array to store timestamps that have data, sorted in ascending order
+  uint256[] private timestamps;
+
+  // Mapping to track the most recent timestamp for each user
+  mapping(address => uint256) private userLastTimestamp;
+
+  // Events
+  event DataUploaded(
+    uint256 timestamp,
+    address indexed uploader,
+    address indexed user,
+    uint256 value
+  );
+
+  constructor(address _owner) AccessControlDefaultAdminRules(0, _owner) {
+    // Grant the deployer the uploader role
+    _grantRole(UPLOADER_ROLE, _owner);
+  }
+
+  /**
+   * @dev Override to prevent the owner from losing their uploader role
+   */
+  function revokeRole(bytes32 role, address account) public virtual override {
+    require(
```

Comment:
we usually do `revert CustomError`.  We had that enabled as a linter option at one point too.  Should we follow the "revert pattern"?